### ISSUE TITLE
Out of memory when loading several thousand records per second with long-running connections

### ISSUE THREAD

**Issue by gaqzi** [cockroachdb/cockroach#41181]
**Describe the problem**

I am evaluating Cockroach as a potential replacement for AWS Postgres Aurora and I’m having a bit of bad luck. It seems I am running out of memory after running my loader scripts for 10-20 minutes. While reading the documentation I saw:

> If Go allocated memory is larger than a few hundred megabytes, you might have encountered a memory leak.

https://www.cockroachlabs.com/docs/stable/cluster-setup-troubleshooting.html#memory-issues

And I’m seeing that Go allocated is several gigs, per the below screenshot.

![Screenshot 2019-09-27 at 16 38 53](https://user-images.githubusercontent.com/18167/65812332-7a35dc80-e1f8-11e9-84e9-b7efdfbefe40.png)

On top of that I’ve also seen warnings about disk stalls in the logs:
```
* FATAL: disk stall detected: unable to sync log files within 30s
*
*
* ERROR: Reported as error 77e719b7380d4f4eb7a3df746d9a2a1d
*
F190927 08:45:54.857685 44288 util/log/clog.go:1275  disk stall detected: unable to sync log files within 30s
```

**To Reproduce**

For my testing scenario, I’m looking at one of our most troublesome tables, which can be characterized by:
- About 1700 new records per second, append-only and written in micro-batches. There are up to 44 different columns that can be set. For any one record about half of the values are not null.
- Our users can query across this in various ways, but the vast majority are doing ordered scans over two timestamps with some filter combinations after that
- In Aurora, we’re working with daily partitioned tables and while it works we’re wondering if we can get better performance with other databases

For testing out Cockroach I looked at the documentation and got myself a 3-node cluster on AWS. I started out with 3x`i3.xlarge` in insecure mode, just to keep my own testing simpler. And I downloaded the latest version of Cockroach as recommended from the download page: `CCL v19.1.4 @ 2019/08/06 15:34:13 (go1.11.6)`

I also set up 6x loader processes that wrote 2000 records as fast as it could, evenly split across all nodes. In my mind, this to both see what kind of performance to expect, but also to simulate a DR scenario where I had to load as fast as possible.

I ran this same tests multiple times with different EC2 instances since I figured maybe I was running wrongly sized nodes. I tried with `i3.2xlarge` and then I switched to the `c5d` type that is mentioned in the documentation as the ones you use for your own internal testing, `c5d.4xlarge` to get enough disk space for one week of data. 

Below are log excerpts from the three machines in my test cluster:

`.123`
```
I190927 12:27:07.263379 346 server/status/runtime.go:500  [n1] runtime stats: 29 GiB RSS, 289 goroutines, 14 GiB/4.3 GiB/20 GiB GO alloc/idle/total, 7.6 GiB/9.5 GiB CGO alloc/total, 35513.8 CGO/sec, 158.7/7.9 %(u/s)time, 0.0 %gc (0x), 36 MiB/24 MiB (r/w)net
W190927 12:27:08.402297 184414 vendor/google.golang.org/grpc/clientconn.go:1304  grpc: addrConn.createTransport failed to connect to {172.21.119.217:26257 0  <nil>}. Err :connection error: desc = "transport: Error while dialing cannot reuse client connection". Reconnecting...
W190927 12:27:08.402332 184414 vendor/google.golang.org/grpc/clientconn.go:1440  grpc: addrConn.transportMonitor exits due to: context canceled
I190927 12:27:12.571320 186817 storage/replica_command.go:247  [n1,split,s1,r396/1:/Table/53/{1/48993…-2/2019-…}] initiating a split of this range at key /Table/53/1/489936756648509443 [r397] (64 MiB above threshold size 64 MiB); r396/3 is being probed (may or may not need a Raft snapshot)
W190927 12:27:13.368628 348 server/node.go:880  [n1,summaries] health alerts detected: {Alerts:[{StoreID:1 Category:METRICS Description:ranges.underreplicated Value:215 XXX_NoUnkeyedLiteral:{} XXX_sizecache:0}] XXX_NoUnkeyedLiteral:{} XXX_sizecache:0}
I190927 12:27:14.148184 186753 storage/replica_command.go:247  [n1,split,s1,r394/1:/{Table/53/3/4…-Max}] initiating a split of this range at key /Table/53/3/489936575284805635/273442200 [r398] (64 MiB above threshold size 64 MiB); r394/3 is being probed (may or may not need a Raft snapshot)
fatal error: runtime: out of memory
```

`.74`
```
W190927 12:27:30.387383 182033 vendor/google.golang.org/grpc/clientconn.go:953  Failed to dial 172.21.126.123:26257: context canceled; please retry.
I190927 12:27:30.560376 275 vendor/github.com/cockroachdb/circuitbreaker/circuitbreaker.go:447  [n3] circuitbreaker: rpc [::]:26257->2 event: BreakerReset
fatal error: runtime: out of memory
```

`.217`
```
W190927 12:30:01.371468 171682 vendor/google.golang.org/grpc/clientconn.go:953  Failed to dial 172.21.119.74:26257: context canceled; please retry.
W190927 12:30:01.327935 171735 storage/replica_range_lease.go:982  [n2,s2,r412/3:/Table/53/{1/48993…-2/2019-…}] have been waiting 1m0s attempting to acquire lease
W190927 12:30:01.329163 144 gossip/gossip.go:1499  [n2] first range unavailable; resolvers exhausted
W190927 12:30:01.376823 171825 server/server_engine_health.go:78  [n2] disk stall detected: unable to write to <no-attributes>=/data/cockroach-data within 10s 

** Compaction Stats [default] **
Level    Files   Size     Score Read(GB)  Rn(GB) Rnp1(GB) Write(GB) Wnew(GB) Moved(GB) W-Amp Rd(MB/s) Wr(MB/s) Comp(sec) Comp(cnt) Avg(sec) KeyIn KeyDrop
----------------------------------------------------------------------------------------------------------------------------------------------------------
  L0      1/0   11.56 MB   0.5      0.0     0.0      0.0       4.1      4.1       0.0   1.0      0.0     48.1        87       371    0.234       0      0
  L4     16/0   63.09 MB   1.0      9.9     3.2      6.6       9.6      3.0       0.0   3.0     29.2     28.5       346       144    2.403    338M   171K
  L5     38/0   124.14 MB   1.0      4.6     2.4      2.3       4.6      2.3       1.3   1.9     36.6     36.1       130       194    0.669    159M   319K
  L6    255/0    1.50 GB   0.0      9.9     3.1      6.8       7.8      1.0       0.5   2.5     42.5     33.5       238       460    0.518    367M    47M
 Sum    310/0    1.69 GB   0.0     24.4     8.8     15.6      26.1     10.5       1.8   6.4     31.2     33.4       801      1169    0.685    865M    47M
 Int      0/0    0.00 KB   0.0      0.2     0.1      0.2       0.2      0.1       0.0  19.4      2.7      2.8        81        10    8.085     11M   1209
Uptime(secs): 8252.4 total, 160.6 interval
Flush(GB): cumulative 4.077, interval 0.011
AddFile(GB): cumulative 0.000, interval 0.000
AddFile(Total Files): cumulative 0, interval 0
AddFile(L0 Files): cumulative 0, interval 0
AddFile(Keys): cumulative 0, interval 0
Cumulative compaction: 26.09 GB write, 3.24 MB/s write, 24.39 GB read, 3.03 MB/s read, 801.0 seconds
Interval compaction: 0.22 GB write, 1.40 MB/s write, 0.22 GB read, 1.37 MB/s read, 80.9 seconds
Stalls(count): 0 level0_slowdown, 0 level0_slowdown_with_compaction, 0 level0_numfiles, 0 level0_numfiles_with_compaction, 0 stop for pending_compaction_bytes, 0 slowdown for pending_compaction_bytes, 0 memtable_compaction, 0 memtable_slowdown, interval 0 total count
W190927 12:30:01.376843 171748 vendor/google.golang.org/grpc/clientconn.go:1304  grpc: addrConn.createTransport failed to connect to {172.21.126.123:26257 0  <nil>}. Err :connection error: desc = "transport: Error while dialing cannot reuse client connection". Reconnecting...
W190927 12:30:01.376854 171748 vendor/google.golang.org/grpc/clientconn.go:1440  grpc: addrConn.transportMonitor exits due to: grpc: the connection is closing
I190927 12:30:01.377268 171923 internal/client/txn.go:618  [n2] async rollback failed: aborted in distSender: context deadline exceeded
I190927 12:30:01.382278 442 vendor/github.com/cockroachdb/circuitbreaker/circuitbreaker.go:322  [n2] circuitbreaker: rpc [::]:26257->3 tripped: failed to connect to n3 at 172.21.119.74:26257: initial connection heartbeat failed: rpc error: code = DeadlineExceeded desc = context deadline exceeded
I190927 12:30:01.382291 442 vendor/github.com/cockroachdb/circuitbreaker/circuitbreaker.go:447  [n2] circuitbreaker: rpc [::]:26257->3 event: BreakerTripped
I190927 12:30:01.382302 442 rpc/nodedialer/nodedialer.go:143  [n2] unable to connect to n3: failed to connect to n3 at 172.21.119.74:26257: initial connection heartbeat failed: rpc error: code = DeadlineExceeded desc = context deadline exceeded
W190927 12:30:01.389311 171691 storage/replica_write.go:199  [n2,s2,r412/3:/Table/53/{1/48993…-2/2019-…}] have been waiting 130.18s for proposing command RequestLease [/Table/53/1/489936571041120259,/Min).
This range is likely unavailable.
Please submit this message at

  https://github.com/cockroachdb/cockroach/issues/new/choose

along with

	https://yourhost:8080/#/reports/range/412

and the following Raft status: {"id":"3","term":13,"vote":"1","commit":303,"lead":"1","raftState":"StateFollower","applied":176,"progress":{},"leadtransferee":"0"}
I190927 12:30:01.410147 458 server/status/runtime.go:500  [n2] runtime stats: 30 GiB RSS, 274 goroutines, 10 GiB/10 GiB/22 GiB GO alloc/idle/total, 7.7 GiB/9.7 GiB CGO alloc/total, 33.0 CGO/sec, 0.6/0.3 %(u/s)time, 0.0 %gc (0x), 18 MiB/287 KiB (r/w)net
W190927 12:30:01.413515 426 storage/store.go:3704  [n2,s2,r37/3:/Table/53/3/4899107{340…-800…}] handle raft ready: 130.6s [processed=0]
W190927 12:30:01.329265 325 storage/store.go:3704  [n2,s2,r186/3:/Table/53/2/2019-09-21T0…] handle raft ready: 132.3s [processed=0]
W190927 12:30:01.329282 430 storage/store.go:3704  [n2,s2,r240/3:/Table/53/2/2019-09-21T1…] handle raft ready: 51.7s [processed=0]
I190927 12:30:01.341529 454 gossip/gossip.go:557  [n2] gossip status (stalled, 3 nodes)
gossip client (0/3 cur/max conns)
gossip server (1/3 cur/max conns, infos 8110/26433 sent/received, bytes 1535643B/3873550B sent/received)
  1: 172.21.126.123:26257 (2m26s)
I190927 12:30:01.413821 454 gossip/gossip.go:557  [n2] gossip status (stalled, 3 nodes)
gossip client (0/3 cur/max conns)
gossip server (1/3 cur/max conns, infos 8110/26433 sent/received, bytes 1535643B/3873550B sent/received)
  1: 172.21.126.123:26257 (2m26s)
W190927 12:30:01.343293 446 storage/replica_range_lease.go:982  [n2,s2,r2/3:/System/NodeLiveness{-Max}] have been waiting 1m0s attempting to acquire lease
```

I was not able to get the log output from the URL above, this server got non-responsive and crashed after a while, and when I finally got in and started Cockroach there was nothing there anymore. 

**Expected behavior**
I was expecting to be able to leave my scripts running for about 22 hours and then have a week worth of data in Cockroach so that I could start testing query performance.

**Additional data / screenshots**

I will send an email to support referencing this issue with table schema and all logs.

**Environment:**
 - CockroachDB version CCL v19.1.4 @ 2019/08/06 15:34:13 (go1.11.6)
 - Server OS: Ubuntu 18.04
- AWS instance type: c5d.4xlarge
- Python through psycopg2 2.8.2
- Servers started by: `cockroach start --insecure --join=172.21.126.123,172.21.119.74,172.21.119.217 --cache=.25 --max-sql-memory=.25 --advertise-addr <ip of server>`

**Comment by gaqzi** [comment]
I couldn't leave this be, and I found something that explained the growth in memory utilization.

With my code change:
![Screenshot 2019-09-28 at 19 44 13](https://user-images.githubusercontent.com/18167/65816403-a1f36780-e22d-11e9-9d72-bf1e649fcde8.png)

Reverting to my previous code on one loader process:
![Screenshot 2019-09-28 at 19 58 13](https://user-images.githubusercontent.com/18167/65816405-a9b30c00-e22d-11e9-8375-d2e1d5dc334b.png)

And after running the loader process for a while and starting to get close to OOM territory, I killed it and got:
![Screenshot 2019-09-28 at 20 07 10](https://user-images.githubusercontent.com/18167/65816432-11695700-e22e-11e9-846e-fbfca31ca47c.png)

The change I introduced was to force a new connection for each batch write (2000 `VALUES (value),(value),…`). This is a bit surprising to me as this is how I keep doing it with Postgres/Aurora, get a connection in the beginning, and then ask for a cursor when you're doing actual work. Because what I have always been told is that getting a connection set up is expensive, so avoid doing that. Admittedly, Postgres specific advice.

My cursor fetch code used to handle both the working (close the connection after each write) and keep the connection around but release the cursor:

```python
@contextmanager
def _get_cursor():
    global CONN
    if CONN is None:
        CONN = _get_conn()

    with CONN.cursor() as cur:
        yield cur

    if not SHARE_CONN:
        CONN.close()
        CONN = None
```

I'll see if I can find a more condensed repro than what I currently have, and if I do, I'll update here.

**Comment by gaqzi** [comment]
Okay, PEBKAC going on here. 😅 

After running all successfully for a while I wanted to do some `SELECT`s to see how everything was going and then I discovered that no data was available. I then discovered that I had to do `conn.commit()` with psycopg2. 

I did while manually testing do `cur.execute('commit')` which raised an exception so I assumed it was in some kind of auto commit mode. And I now realize that when I manually checked the table it must've had some record from another manual test. Urgh. This is where I curse myself for not TDDing this thing despite it trivial spike thing, TDD is a lot better at making sure I don't mix up my states. :) 

Sorry to anyone else that ends up reading this. :)
